File: Python/Scrapping_Scripts/Recherche/Bienici_Recherche.py
import requests
import json
from datetime import datetime, timedelta

def get_old_data(filename):
    try:
        with open(filename, 'r', encoding='utf-8') as file:
            return json.load(file)
    except FileNotFoundError:
        return []
    
BASE_URL = 'https://www.bienici.com/realEstateAds.json'
PARAMS = {
    'filters': json.dumps({
        "size": 24,
        "from": 0,
        "showAllModels": False,
        "filterType": "rent",
        "propertyType": ["house", "flat", "loft", "castle", "townhouse"],
        "page": 1,
        "sortBy": "relevance",
        "sortOrder": "desc",
        "onTheMarket": [True],
        "zoneIdsByTypes": {"zoneIds": ["-7444"]}
    }),
    'extensionType': 'extendedIfNoResult',
    'access_token': 'XFCgRiIm/5gxwcxQg0TU81XoA5F0IuwnwxhoHtf5VDI=:653a380e4d571100b2ac40e6',
    'id': '653a380e4d571100b2ac40e6'
}

all_data = []
page_number = 1 

while True:
    print(f"Fetching page {page_number}...")  # Print the current page number
    
    response = requests.get(BASE_URL, params=PARAMS)
    if response.status_code == 200:
        data = response.json()
        
        # Check if the returned data is less than the expected size.
        if not data['realEstateAds'] or len(data['realEstateAds']) < 24:
            all_data.extend(data['realEstateAds'])  # add remaining ads
            break
            
        all_data.extend(data['realEstateAds'])
        PARAMS['filters'] = json.dumps({**json.loads(PARAMS['filters']), "from": len(all_data), "page": int(len(all_data)/24) + 1})
        page_number += 1  # Increment the page number for next iteration
    else:
        print(f"Erreur {response.status_code}: {response.text}")
        break


#Obtenir la date actuelle sous forme de chaîne
current_date_string = datetime.now().strftime('%Y-%m-%d')

# Chargement des anciennes données
previous_day_string = (datetime.now() - timedelta(days=1)).strftime('%Y-%m-%d')
old_file_name = f'../../Resultat_Recherche/Bienici_Recherche/Data_Bienici_Recherche_{previous_day_string}.json'
old_data = get_old_data(old_file_name)

# Identifier les nouvelles annonces et les annonces supprimées
new_announcements = [item for item in all_data if item['id'] not in [old_item['id'] for old_item in old_data]]
removed_announcements = [item for item in old_data if item['id'] not in [new_item['id'] for new_item in all_data]]

# Créez la liste des annonces à jour en excluant les annonces supprimées
updated_data = [item for item in all_data if item['id'] not in [removed_item['id'] for removed_item in removed_announcements]]

# Sauvegardez toutes les annonces scrappées du jour
output_file_name = f'../../Resultat_Recherche/Bienici_Recherche/Data_Bienici_Recherche_{current_date_string}.json'
with open(output_file_name, 'w', encoding='utf-8') as file:
    json.dump(all_data, file, ensure_ascii=False, indent=4)

# Sauvegardez les annonces mises à jour
updated_file_name = f'../../Resultat_Recherche/Up_To_Date_Recherche/Updated_Data_Bienici_Recherche_{current_date_string}.json'
with open(updated_file_name, 'w', encoding='utf-8') as file:
    json.dump(updated_data, file, ensure_ascii=False, indent=4)

print(f"Total d'annonces scrappées: {len(all_data)}")
print(f"Données du jour sauvegardées dans {output_file_name}!")
print(f"{len(new_announcements)} nouvelle(s) annonce(s).")
print(f"{len(removed_announcements)} annonce(s) supprimée(s).")
print(f"{len(updated_data) - len(new_announcements)} annonce(s) conservée(s).")
print(f"Données mises à jour sauvegardées dans {updated_file_name}!")

File: Javascript/Scrapping_Scripts/Recherche/BienIci_Recherche.js
const axios = require('axios');
const fs = require('fs');
const path = require('path');
const { getCurrentDateString, getPreviousDateString } = require('../dateUtils');
const { getOldData } = require('../dataUtils');

const BASE_URL = 'https://www.bienici.com/realEstateAds.json';

const cities = [
    { name: "Paris", ids: ["-7444"]},
    { name: "Lyon", ids: ["-120965"] },
    { name: "Villeurbanne", ids: ["-120955"] },
    { name: "Vénissieux", ids: ["-164210"] },
    { name: "Saint-Etienne", ids: ["-117905"] },
    { name: "Marseille", ids: ["-76469"] },
    { name: "Toulouse", ids: ["-35738"] },
    { name: "Bordeaux", ids: ["-105270"] },
    { name: "Nantes", ids: ["-59874"] },
    { name: "Rennes", ids: ["-54517"] },
    { name: "Lille", ids: ["-58404"] },
    { name: "Angers", ids: ["-178351"] },
    { name: "Grenoble", ids: ["-80348"] },
    { name: "Montreuil", ids: ["-129423"] },
    { name: "Cergy", ids: ["-120955"] },
];

const delay = duration => new Promise(resolve => setTimeout(resolve, duration));

const fetchDataForCity = async (city) => {
    let allData = [];
    let pageNumber = 1;
    let PARAMS = {
        filters: JSON.stringify({
            size: 24,
            from: 0,
            showAllModels: false,
            filterType: "rent",
            propertyType: ["house", "flat", "loft", "castle", "townhouse"],
            page: 1,
            sortBy: "relevance",
            sortOrder: "desc",
            onTheMarket: [true],
            zoneIdsByTypes: { "zoneIds": city.ids }
        }),
        extensionType: 'extendedIfNoResult',
        access_token: 'XFCgRiIm/5gxwcxQg0TU81XoA5F0IuwnwxhoHtf5VDI=:653a380e4d571100b2ac40e6',
        id: '653a380e4d571100b2ac40e6'
    };

    while (true) {
        console.log(`Fetching page ${pageNumber} for ${city.name}...`);

        try {
            const response = await axios.get(BASE_URL, { params: PARAMS });

            if (response.status === 200) {
                const data = response.data;

                if (!data.realEstateAds || data.realEstateAds.length < 24) {
                    allData.push(...data.realEstateAds);
                    break;
                }

                allData.push(...data.realEstateAds);
                PARAMS.filters = JSON.stringify({ ...JSON.parse(PARAMS.filters), "from": allData.length, "page": Math.floor(allData.length / 24) + 1 });
                pageNumber++;
            } else {
                console.log(`Error ${response.status}: ${response.statusText}`);
                break;
            }
        } catch (error) {
            console.error('An error occurred:', error);
            break;
        }
    }

    const currentDateString = getCurrentDateString();
    const previousDateString = getPreviousDateString();
    const oldFileName = path.join(__dirname, `../../Resultat_Recherche/Bienici_Recherche/Data_Bienici_Recherche_${city.name}_${previousDateString}.json`);
    const oldData = getOldData(oldFileName);

    const newAnnouncements = allData.filter(item => !oldData.some(oldItem => oldItem.id === item.id));
    const removedAnnouncements = oldData.filter(item => !allData.some(newItem => newItem.id === item.id));
    const updatedData = allData.filter(item => !removedAnnouncements.some(removedItem => removedItem.id === item.id));

    const outputFileName = path.join(__dirname, `../../Resultat_Recherche/Bienici_Recherche/Data_Bienici_Recherche_${city.name}_${currentDateString}.json`);
    const updatedFileName = path.join(__dirname, `../../Resultat_Recherche/Up_To_Date_Recherche/BienIci_Recherche_Up_To_Date/Updated_Data_Bienici_Recherche_${city.name}_${currentDateString}.json`);

    fs.writeFileSync(outputFileName, JSON.stringify(allData, null, 4), 'utf-8');
    fs.writeFileSync(updatedFileName, JSON.stringify(updatedData, null, 4), 'utf-8');

    console.log(`Total d'annonces scrappées pour ${city.name}: ${allData.length}`);
    console.log(`Données du jour pour ${city.name} sauvegardées dans ${outputFileName}`);
    console.log(`TOTAL_NOUVELLES_ANNONCES pour ${city.name}: ${newAnnouncements.length} nouvelles annonces sur BienIci.`);
    console.log(`${removedAnnouncements.length} annonce(s) supprimée(s) pour ${city.name}.`);
    console.log(`${updatedData.length - newAnnouncements.length} annonce(s) conservée(s) pour ${city.name}.`);
    console.log(`Données mises à jour pour ${city.name} sauvegardées dans ${updatedFileName}`);
};

const fetchAllCities = async () => {
    for (const city of cities) {
        await fetchDataForCity(city);
        await delay(60000); // Délai de 1 minute
    }
};

fetchAllCities();
